conda_recipe_manager/parser/dependency.py

```diff
@@ -0,0 +1,108 @@
+    Converts a dependency section enumeration to the equivalent string found in the recipe, based on the current
+    schema.
+    :param section: Target dependency section
+    :param schema: Target recipe schema
+    :returns: String equivalent of the recipe schema
+    """
+    # `match` is used here so the static analyzer can ensure all cases are covered
+    match schema:
+        case SchemaVersion.V0:
+            match section:
+                case DependencySection.BUILD:
+                    return "build"
+                case DependencySection.HOST:
+                    return "host"
+                case DependencySection.RUN:
+                    return "run"
+                case DependencySection.RUN_CONSTRAINTS:
+                    return "run_constrained"
+                case DependencySection.RUN_EXPORTS:
+                    return "run_exports"
+        case SchemaVersion.V1:
+            match section:
+                case DependencySection.BUILD:
+                    return "build"
+                case DependencySection.HOST:
+                    return "host"
+                case DependencySection.RUN:
+                    return "run"
+                case DependencySection.RUN_CONSTRAINTS:
+                    return "run_constraints"
+                case DependencySection.RUN_EXPORTS:
+                    return "run_exports"
+
+
+def str_to_dependency_section(s: str) -> Optional[DependencySection]:
+    """
+    Converts a dependency section string to a section enumeration.
+    :param s: Target string to convert
+    :returns: String equivalent of the recipe schema. None if the string is unrecognized.
+    """
+    # `match` is used here so the static analyzer can ensure all cases are covered
+    match s.strip().lower():
+        case "build":
+            return DependencySection.BUILD
+        case "host":
+            return DependencySection.HOST
+        case "run":
+            return DependencySection.RUN
+        case "run_constrained":  # V0
+            return DependencySection.RUN_CONSTRAINTS
+        case "run_constraints":  # V1
+            return DependencySection.RUN_CONSTRAINTS
+        case "run_exports":
+            return DependencySection.RUN_EXPORTS
+        case _:
+            return None
+
+
+class Dependency(NamedTuple):
+    """
+    Structure that contains metadata about a dependency found in the recipe. This is immutable by design.
+    """
+
+    # Owning package name
+    required_by: str
+    # Path in the recipe where this dependency was found
+    path: str
+    # Identifies what kind of dependency this is
+    type: DependencySection
+    # Parses the dependency's name and version constraints
+    match_spec: MatchSpec
+    # The selector applied to this dependency, if applicable
+    selector: Optional[SelectorParser] = None
+
+
+# Maps-out dependencies found in a recipe. Maps package name -> list of parsed dependencies.
+DependencyMap = dict[str, list[Dependency]]
```

Comment:
This structure contains a list of dependencies for the sake of simplicity. I figured for most of our use cases, we will end up iterating over all the dependencies anyways. At the very least, we'll likely need to filter one or more `DependencySection`s at a time from this list.

So I thought a `dict[str, dict[DependencySection, list[Dependency]]` would just end up causing more headaches and would not buy us much in terms of performance. Also I haven't seen a recipe file with thousands of listed dependencies so the search space is small.